MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -265,74 +265,290 @@ SetUefiImageMemoryAttributes (
 extern LIST_ENTRY  mGcdMemorySpaceMap;
 
 // ---------------------------------------
-//         SPECIAL REGION LOGIC
+//     LINKED LIST SUPPORT FUNCTIONS
 // ---------------------------------------
 
 /**
-  Sorts the MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY list by Start
+  Inserts the input SpecialRegionToInsertLink into SpecialRegionList based on the
+  SpecialRegionToInsert.SpecialRegion.Start field.
 
-  @param[in] SpecialRegionList Head of the list to be sorted
+  @param[in] SpecialRegionList           Pointer to the head of the MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY list
+  @param[in] SpecialRegionToInsertLink   Pointer to the list entry of the MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY to insert
+
+  @retval EFI_SUCCESS                    MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY inserted into the list
+  @retval EFI_INVALID_PARAMETER          CodeSectionList or CodeSectionToInsertLink were NULL
 **/
 STATIC
-VOID
-SortMemoryProtectionSpecialRegionList (
-  IN LIST_ENTRY  *SpecialRegionList
+EFI_STATUS
+OrderedInsertSpecialRegionListEntry (

```

Comment:
does this need to handle the case of the SpecialRegionToInsertLink having overlap with the existing list?  Right now it will just create a new entry even if there is overlap.

Reply:
For the linked list function itself. I don't think it needs to handle the scenario you describe.

For the logic that's actually adding the region entry, it might be a good redundancy measure but it's not immediately obvious what the best course of action is when there is overlap. Maybe adding a remove() function to the protocol is a good idea so the caller of add() can remove the overlapping entry before calling add() again, or maybe the best way of resolving it would be to just have whatever is being inserted "overwrite" the existing special entry region which might necessitate one special region list entry becoming three.